Feature/206838 support lists of documents

Ethan-Lambourne opened the pull request:
In this PR:
- Created the `TprDocuments` tag helper along with its children tag helpers and context file;
- Added styles for the new tag helper in the `_tpr-documents.scss` file to mimic the current styles on the corporate website;
- Created document types for `TPRDocuments `and `TPRDocument ` along with their relative settings document types;
- Added element type and property aliases to `Tpr.Frontend.Umbraco.ElementTypeAliases` and `Tpr.Frontend.Umbraco.PropertyAliases` respectively;
- Created `TPRDocuments.cshtml` view using the tag helpers to render the html using properties gathered from Umbraco;
- Added support for `TPR Documents` on the `TPR Block Grid` and `TPR Block List`;
- Created a backoffice view for `TPR Documents` using AngularJS;
- Created an example page displaying a use case of the `TPR Documents` block and added a link to it from the homepage;
- Created markdown documentation for the new feature in `tpr-documents.md` and linked to it from `README.md`;
- Incremented the version of packages to v7.4.4.

Once this PR is approved [a new version of all packages should be published](https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions/blob/develop/docs/contributing/publish-to-nuget.md).

Commit:
Tag helpers created (WIP)

Commit:
Adding umbraco support (WIP)

Commit:
Fully implemented and supported in umbraco

Commit:
custom backoffice view

Commit:
clean export from umbraco exampleapp, incrementing package version, moving files to TPR.Frontend.Umbraco

Ethan-Lambourne committed:
Create tpr-documents.md

will add images and link to page from README.md in a moment

Ethan-Lambourne committed:
Uploading images for tpr-documents.md

Ethan-Lambourne committed:
Update tpr-documents.md

Ethan-Lambourne committed:
Update tpr-documents.md

Ethan-Lambourne committed:
Update README.md

added link to tpr-documents.md

Commit:
catching up with develop

Commit:
Merge branch 'feature/206838-Support-lists-of-documents' of https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions into feature/206838-Support-lists-of-documents

Commit:
getting rid of unneeded duplicates

Commit:
Updating alias names

Commit:
fixing date on documents example page

villarin reviewed (changes requested):
Hi @Ethan-Lambourne 

PR reviewed locally and there is only one requests from me to change the published date field in [the tprdocument document type](https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions/pull/409/files#diff-016096fa41a32d02f089cfbbaeb7f5e1268ff182d0ac976c205b03db2e5a52bb) to date picker and normalise how the date field is rendered in the frontend. 

All other checks passed succesfully:

- Documentation for component has been added
- Example page has been created
- Block grid component working as expected.

villarin commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/ContentTypes/tprdocument.config:
```diff
@@ -0,0 +1,114 @@
+      <Name>Date published</Name>
+      <Alias>datePublished</Alias>
+      <Definition>0cc0eba1-9960-42c9-bf9b-60e150b429ae</Definition>
+      <Type>Umbraco.TextBox</Type>
```
Please change type to date picker and normalise the date value rendering on the relevant backoffice app plugin and block view.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

villarin replied:
Tested locally and working as expected!

sussexrick commented on Directory.Build.props:
```diff
@@ -1,5 +1,5 @@
 <Project>
 	<PropertyGroup>
-        <Version>7.3.4</Version>
```
I would suggest removing this change because this will actually go into 8.something due to other changes, but you weren't to know that.

More importantly for your learning, this would've been 7.4.0 because the last number resets when you increment the middle one. Read about [semantic versioning](https://semver.org/).

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
Out-of-date again unfortunately. Maybe just take it out of this PR and we'll do a separate small PR with a version bump when we're ready to go.

villarin replied:
Ideally get rid of the file change on this branch or sync it up with develop

Ethan-Lambourne replied:
Release version is currently synced with develop.

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/Content/home.config:
```diff
@@ -40,7 +40,7 @@
       "udi": "umb://element/ca521a0028b74cdf939fa9c310a4fc76",
       "text": {
-        "markup": "<p>This text is using <a href=\"https://design-system.service.gov.uk/styles/typography/\">GOV.UK Design System typography</a>.</p>\n<p>This example application demonstrates GOV.UK Design System components and styling. By default an extra layer of styling is applied to demonstrate the brand of The Pensions Regulator (TPR), but this can be switched off in appsettings.json to retu ...
+        "markup": "<p>This text is using <a href=\"https://design-system.service.gov.uk/styles/typography/\">GOV.UK Design System typography</a>.</p>\n<p>This example application demonstrates GOV.UK Design System components and styling. By default an extra layer of styling is applied to demonstrate the brand of The Pensions Regulator (TPR), but this can be switched off in appsettings.json to retu ...
```
This is a TPR component, not a GOV.UK one, so it should be in the list at the bottom of the page. Just to warn you, you could get a merge conflict because #408 also updates this list to make sure TPR components are in the TPR list.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
Your demo page isn't listed on the home page at all now.

villarin replied:
Can see the documents demo page available as below 
![image](https://github.com/user-attachments/assets/2cefcced-8d46-480c-8aee-136771d69fb1)

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/Media/testpdf.config:
```diff
@@ -0,0 +1,17 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Media Key="709a03ab-4635-4320-a1da-ebe21cd650b6" Alias="Testpdf" Level="1">
```
You're including the Umbraco metadata for this file and you link to it from the example page, but you haven't included the file itself so the link is broken.

sussexrick replied:
Still an issue.

villarin replied:
Files still missing @Ethan-Lambourne 
![image](https://github.com/user-attachments/assets/b50ba590-8f85-40be-be77-a52356d501aa)

Ethan-Lambourne replied:
They should be included now - I forgot to exclude them from the gitignore

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/Content/documents.config:
```diff
@@ -0,0 +1,173 @@
+      "columnSize": "",
+      "columnSizeFromDesktop": "",
+      "contentTypeKey": "d66ffc80-0d15-4056-8acc-cdc8a7e6ce1b",
+      "cssClasses": "govuk-body",
```
This class might be left in from a test, but it results in `govuk-body govuk-body` in the final HTML.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
You render `govuk-list` correctly instead, but you haven't updated the example in documentation or the backoffice preview.

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/Views/Shared/TPR/TPRDocuments.cshtml:
```diff
@@ -0,0 +1,56 @@
+@addTagHelper *, ThePensionsRegulator.Frontend
+
+@{
+    var OuterCssClasses = Model.Settings.Value<string>(GovUkPropertyAliases.CssClasses);
```
variables should be `camelCase`

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/Views/Shared/TPR/TPRDocuments.cshtml:
```diff
@@ -0,0 +1,56 @@
+@using Umbraco.Cms.Core.Models.Blocks
+@using Umbraco.Cms.Web.Common
+@using GovUkPropertyAliases = GovUk.Frontend.Umbraco.PropertyAliases
+@using TprPropertyAliases = ThePensionsRegulator.Frontend.Umbraco.TprPropertyAliases
```
A recent change means you don't need to alias these namespaces any more. It can just be a standard using statement, if you need it at all.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/Views/Shared/TPR/TPRDocuments.cshtml:
```diff
@@ -0,0 +1,56 @@
+    var documents = Model.Content.Value<IEnumerable<BlockListItem>>(TprPropertyAliases.DocumentsBlockList);
+}
+
+<tpr-documents class="tpr-documents govuk-body @OuterCssClasses">
```
`govuk-body` should be `govuk-list`, and I would expect the tag helper always to add both `tpr-documents` and `govuk-list` internally, not to expect the consuming page to remember to add them.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
You render `govuk-list` correctly now, but you haven't updated the example in documentation or the backoffice preview.

villarin replied:
still needs work - check the component preview as suggested by @sussexrick

Ethan-Lambourne replied:
Just updated the backoffice preview class from govuk-body to govuk-list

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/Views/Shared/TPR/TPRDocuments.cshtml:
```diff
@@ -0,0 +1,56 @@
+            var docKB = document.Content.Value<string>(TprPropertyAliases.DocumentFileSize);
+            var docPages = document.Content.Value<string>(TprPropertyAliases.DocumentNumberOfPages);
+
+            <tpr-document href="@docHref" kbsize="@docKB" pages="@docPages" class="tpr-document @innerCssClasses">
```
I would expect the tag helper always to add `tpr-document` internally, not to expect the consuming page to remember to add them

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/Views/Shared/TPR/TPRDocuments.cshtml:
```diff
@@ -0,0 +1,56 @@
+    @foreach (var document in documents)
+    {
+        var innerCssClasses = document.Settings.Value<string>(GovUkPropertyAliases.CssClasses);
+        var docLinkType = document.Content.Value<Link>(TprPropertyAliases.Document).Type;
```
What if `document.Content.Value<Link>(TprPropertyAliases.Document)` returns `null`? This view throws an error:

> RuntimeBinderException: Cannot perform runtime binding on a null reference

It is a required field, but if you link to a page (or probably a document too) and then the page is deleted this can still return `null`. This view should cope with that possibility, probably by not rendering the document at all. Not much point rendering the document details without the link.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
This is still a problem, and I also found that it occurs if the link points to an unpublished page.

You need to get the value as a nullable `Link` and then check it for null before you use it:

```csharp
var link = document.Content.Value<Link?>(TprPropertyAliases.Document);
if (link is not null) {
    /// now it's safe to use
}
```

villarin replied:
still many duplicate references to document.Content.Value<Link?> here - please rewrite as suggested by @sussexrick - you can also use the strongly typed models if it makes it easier

Ethan-Lambourne replied:
This has been implemented now, I also simplified the variables a bit to make the partial view more readable

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/Views/Shared/TPR/TPRDocuments.cshtml:
```diff
@@ -0,0 +1,56 @@
+
+                <tpr-document-title>@docTitle</tpr-document-title>
+
+                <dd>Published: @docDatePublished</dd>
```
This should be a tag helper rather than plain HTML.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-documents.scss:
```diff
@@ -0,0 +1,29 @@
+.tpr-documents {
+    margin-bottom: 2rem;
```
Use the standard `margin-bottom` spec for all components, which is `@include govuk-responsive-margin(6, "bottom");`

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-documents.scss:
```diff
@@ -0,0 +1,29 @@
+}
+
+.tpr-document .fileicon.pdf {
+    background-color: #a91717;
```
Use TPR colours for these, referencing them by TPR colour variables. `$tpr-colour-royal-blue` works for Word, `$tpr-colour-fire-brick` works for PDF and `$tpr-colour-camarone` will work for Excel.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
There's nothing in the SCSS for Excel.

Ethan-Lambourne replied:
Colours for PDF, Word, Powerpoint, and Excel type files have been added

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/Views/Shared/TPR/TPRDocuments.cshtml:
```diff
@@ -0,0 +1,56 @@
+            var docKB = document.Content.Value<string>(TprPropertyAliases.DocumentFileSize);
+            var docPages = document.Content.Value<string>(TprPropertyAliases.DocumentNumberOfPages);
+
+            <tpr-document href="@docHref" kbsize="@docKB" pages="@docPages" class="tpr-document @innerCssClasses">
```
Don't ask the user for the file size - you can get the media item from Umbraco and find that out automatically.

You can remove that field from the document type (which in turn means you need to update your screenshot and your text in a few places).

sussexrick replied:
Still to do.

Ethan-Lambourne replied:
The partial view now automatically finds the umbracoBytes, converts it to KB, and plugs that into the tag helper. Also updated the screenshots in the docs to reflect changes

villarin replied:
umbracoBytes coming up as 0 - so you may need an additional call to some extension like Model.MediaById(documentLink.Content.Id).UmbracoBytes to obtain the correct size

villarin replied:
HI @Ethan-Lambourne 
As discussed earlier I have successfully replicated the automatic media size calculation via the umbracoBytes field.
This works after a clean usync import from a vanilla umbraco database.
So resolving and merging this now

sussexrick commented on ThePensionsRegulator.Frontend/TagHelpers/TprDocumentTitleTagHelper.cs:
```diff
@@ -0,0 +1,34 @@
+            documentContext.DocumentTitle = await output.GetChildContentAsync();
+
+            output.PreElement.SetHtmlContent("<dt>");
+            output.TagName = $"a href='{documentContext.Href}'";
```
This link (and all links) should use the CSS class `govuk-link`. Don't forget to update examples in your docs and the backoffice preview.

sussexrick replied:
Docs and backoffice preview not updated.

Ethan-Lambourne replied:
The code generated by the tag helpers as well as the docs and backoffice preview now use the govuk-link class

sussexrick commented on ThePensionsRegulator.Frontend/TagHelpers/TprDocumentTitleTagHelper.cs:
```diff
@@ -0,0 +1,34 @@
+            }
+            else if (documentContext.Href!.EndsWith(".docx"))
+            {
+                output.PostContent.SetHtmlContent($"<br><span class='doc fileicon'>WORD</span> {documentContext.KbSize}KB, {documentContext.Pages} pages </dt>");
```
"WORD" should be "Word", and you can use CSS `text-transform: uppercase` to present it the way you want. That's because if you put it in all-caps a screen-reader is likely to read it out like an abbreviation, letter-by-letter. You want that for "P. D. F." but not for "W. O. R. D."

sussexrick commented on docs/components/tpr-documents.md:
```diff
@@ -0,0 +1,40 @@
+    <dd>This is a test PDF, uploaded for example purposes.</dd>
+  </div>
+  <div class="tpr-document ">
+    <dt><a href='/'>Home page</a href='/'></dt>
```
Remove `href` from your closing tags (in both examples)

sussexrick replied:
Not done.

Ethan-Lambourne replied:
Docs have been updated and duplicate tags have been removed

sussexrick commented on docs/components/tpr-documents.md:
```diff
@@ -0,0 +1,40 @@
+
+![TPR Documents custom inner block list](../images/tpr-documents-inner-blocklist.png)
+
+After pressing 'Add TPR Document' the following menu should appear, prompting you to link the documet as well as fill out any relevant information about the document. The 'File size' and 'Number of pages' fields should only be filled out if the document you are linking to is a piece of media such as a .PDF or .DOCX file.
```
typo: "documet"

sussexrick replied:
Not done

Ethan-Lambourne replied:
Docs have been updated and typo has been fixed

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/ThePensionsRegulator.Frontend.Umbraco.csproj:
```diff
@@ -66,11 +67,17 @@
 	<ItemGroup>
 		<DataTypes Include="..\GovUk.Frontend.Umbraco.ExampleApp\uSync\v9\DataTypes\TPR*.config"></DataTypes>
 		<ContentTypes Include="..\GovUk.Frontend.Umbraco.ExampleApp\uSync\v9\ContentTypes\tpr*.config"></ContentTypes>
+		<None Remove="..\GovUk.Frontend.Umbraco.ExampleApp\uSync\v9\DataTypes\TPRBlockGrid.config" />
```
You don't need to include these two lines or the four below, because wildcards are already including all datatypes and content types that start with "tpr".

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
I think you need to pull the latest `develop` into your branch. Currently you're adding this:

```xml
	<ItemGroup>
	  <Folder Include="uSync\v9\ContentTypes\" />
	  <Folder Include="uSync\v9\DataTypes\" />
	</ItemGroup>
```

...and you're also reverting the backoffice package to an older version.

Ethan-Lambourne replied:
I've removed this now. Weirdly, even when pulling develop into this branch it wasn't removing those unnecessary inclusions

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/App_Plugins/ThePensionsRegulator.Frontend.Umbraco/blocks/views/TPRDocuments.html:
```diff
@@ -0,0 +1,34 @@
+            <dd>Published: {{doc.datePublished}}</dd>
+            <dd ng-bind="doc.description"></dd>
+        </div>
+        <div ng-if="doc.document[0].url.endsWith('.docx')">
```
You also need to handle .dotx, .xlsx, .xlst, and probably .pptx and .csv too. I would recommend speaking with content designers to find out what file types they upload to the public website.

sussexrick replied:
Still to do.

Ethan-Lambourne replied:
Backoffice preview now handles all the required file types

sussexrick commented on ThePensionsRegulator.Frontend/TagHelpers/TprDocumentTitleTagHelper.cs:
```diff
@@ -0,0 +1,34 @@
+            {
+                output.PostContent.SetHtmlContent($"<br><span class='pdf fileicon'>PDF</span> {documentContext.KbSize}KB, {documentContext.Pages} pages </dt>");
+            }
+            else if (documentContext.Href!.EndsWith(".docx"))
```
You also need to handle .dotx, .xlsx, .xlst, and probably .pptx and .csv too. I would recommend speaking with content designers to find out what file types they upload to the public website.

sussexrick replied:
Still to do.

Ethan-Lambourne replied:
The tag helper handles all of the required file types now

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/App_Plugins/ThePensionsRegulator.Frontend.Umbraco/blocks/views/TPRDocuments.html:
```diff
@@ -0,0 +1,34 @@
+﻿
+<dl class="tpr-documents {{block.settingsData.cssClasses}}">
```
The backoffice preview does not line up:

<img width="341" alt="image" src="https://github.com/user-attachments/assets/ec3de281-2156-4b68-88df-b0e0d88e2928" />

This is because you need to apply the `backoffice-block-view` class to the `dl` element for this preview only.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on ThePensionsRegulator.Frontend/TagHelpers/TprDocumentTitleTagHelper.cs:
```diff
@@ -0,0 +1,34 @@
+
+            if (documentContext.Href!.EndsWith(".pdf"))
+            {
+                output.PostContent.SetHtmlContent($"<br><span class='pdf fileicon'>PDF</span> {documentContext.KbSize}KB, {documentContext.Pages} pages </dt>");
```
While `<br>` and attributes with single quotes are valid HTML, they're not consistent with the rest of our code, so `<br />` and double-quoted attributes would be better

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
Example in docs not updated.

Ethan-Lambourne replied:
Example in docs have been updated

sussexrick commented on ThePensionsRegulator.Frontend/TagHelpers/TprDocumentTitleTagHelper.cs:
```diff
@@ -0,0 +1,34 @@
+            output.PreElement.SetHtmlContent("<dt>");
+            output.TagName = $"a href='{documentContext.Href}'";
+
+            if (documentContext.Href!.EndsWith(".pdf"))
```
Using the null-forgiving operator `!` should be a clue to check the null behaviour. If the `href` attribute is not set on the tag helper then it throws:

> NullReferenceException: Object reference not set to an instance of an object.

This could be handled more gracefully, by setting the property as `required` and non-nullable and throwing an informative error when it's missing stating that it's required, or by rendering the component without a link.

![image](https://github.com/user-attachments/assets/d744d611-07d3-4855-887b-1608e93091c6)

sussexrick replied:
Still to do.

Ethan-Lambourne replied:
Tag helper now checks for a valid href and throws an error if it is absent

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/ContentTypes/tprdocument.config:
```diff
@@ -0,0 +1,114 @@
+    </GenericProperty>
+    <GenericProperty>
+      <Key>92ca10a9-d770-471e-bce2-5bb873abdabb</Key>
+      <Name>Description</Name>
```
Check with content designers whether they need any rich text formatting in this description. There _may_ be a need for limited formatting like bold, italic and bulleted lists. 

_If_ there is a need, create a dedicated rich text data type for this component and use that for the property. It would change your HTML to `<dd><p class="govuk-body">...</p></dd>` and that's fine, as `p` is allowed within `dd`.

Ethan-Lambourne replied:
I checked with Chris Mason and he has confirmed that the description should just be regular text, so there is no need for additional formatting:

![image](https://github.com/user-attachments/assets/44733d20-0979-47ca-99f9-8897d8c80273)

sussexrick commented on docs/images/tpr-documents-input-document-form.png:
If you add a document without filling in any of the optional properties the rendered HTML still includes:

```html
<dd>Published:</dd>
<dd></dd>
```

None of that should be there.

sussexrick replied:
Still an issue:

```html
<dd>Published: January 0001</dd>
<dd></dd>
```

![Showing January 0001 as the published date](https://github.com/user-attachments/assets/91b0a5b8-428f-4101-a1bc-ddbdb06459a5)

Ethan-Lambourne replied:
The tag helpers shouldn't be generating any unnecessary html now

sussexrick commented on ThePensionsRegulator.Frontend.Umbraco/Views/Shared/TPR/TPRDocuments.cshtml:
```diff
@@ -0,0 +1,56 @@
+        {
+            var docHref = document.Content.Value<Link>(TprPropertyAliases.Document).Url;
+
+            <tpr-document href="@docHref" class="tpr-document @innerCssClasses">
```
What if they link to a document on someone else's website, eg Pension Protection Fund or Money and Pension Service? You can't _stop_ them adding the number of pages, so why not pass it into the tag helper to be rendered?

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/ContentTypes/tprdocument.config:
```diff
@@ -0,0 +1,114 @@
+      <Type>Umbraco.Integer</Type>
+      <Mandatory>false</Mandatory>
+      <Validation></Validation>
+      <Description><![CDATA[Only use if a file is linked.]]></Description>
```
What if they link to a document on someone else's website, eg Pension Protection Fund or Money and Pension Service? They may want to state the number of pages. I would suggest removing this help text and leaving it up to the user to decide.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/ContentTypes/tprdocuments.config:
```diff
@@ -0,0 +1,50 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<ContentType Key="f08c68b1-f171-4941-83b2-1ededfa019ab" Alias="tprDocuments" Level="2">
+  <Info>
+    <Name>TPR Documents</Name>
```
To be consistent with other components the display name of this shouldn't start with "TPR", it should be "Documents" or "Document list". None of the others that show up in the list of components to add to a page start with "TPR". That would then mean removing "TPR" from the display names of the other new content types, to match this one. 

It's a TPR component but they're not necessarily TPR documents, they might link out to other websites like PPF. They're not necessarily relevant either, though you would hope so. 😂 

The aliases _must_ continue to start with `tpr` though, as that includes them in the package and causes them to render correctly.

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick replied:
Screenshots not updated in the documentation.

villarin replied:
Docs appear updated on the branch now 
![image](https://github.com/user-attachments/assets/f36054b9-f5a0-42a0-be2b-961eeb5df178)

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/DataTypes/TPRBlockGrid.config:
```diff
@@ -1332,6 +1332,27 @@
+      "groupKey": "fe4459ab-dada-4c2f-97f2-1b70aee27310",
+      "iconColor": null,
+      "inlineEditing": false,
+      "label": "",
```
A nice touch here would be to use a [label template](https://docs.umbraco.com/umbraco-cms/10.latest/fundamentals/backoffice/property-editors/built-in-umbraco-property-editors/block-editor/label-property-configuration) that shows the filename in the backoffice UI.

Instead of this:

![image](https://github.com/user-attachments/assets/23e1db2d-dfe8-4aef-93b3-04ca26f62cb4)

It could look more like this:

![image](https://github.com/user-attachments/assets/6f230d99-38a9-4985-8cf8-a29a78457555)

Ethan-Lambourne replied:
I have actioned this comment in a recent commit.

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/Views/TPRDocuments.cshtml:
Nice job on the example page, but it would be good to see one in the other example app too please.

sussexrick replied:
Still to do.

villarin replied:
Available on vanilla example app now 
![image](https://github.com/user-attachments/assets/b359bac2-faad-434f-a510-fb587fceada6)

sussexrick reviewed (changes requested):
Really nice work Ethan, this is well done.

I've made a lot of comments but they're all just tidying up fine details of consistency and edge cases. You've got all the hard stuff spot-on.

Commit:
Responding to comments pt.1

Commit:
merging develop into current branch

Commit:
Responding to comments pt.2

Commit:
merging develop into current branch

Commit:
leftover files

Commit:
Revert "Responding to comments pt.2"

This reverts commit 8abc93ca156f2af4f31f11a9aee1785e7da690e3.

Commit:
Reapply "Responding to comments pt.2"

This reverts commit 97f791ddc8682101ac38c0dd16b79499e15e23ab.

Commit:
reverting to before I pulled from develop

sussexrick reviewed (dismissed):
Following your most recent changes, it is now no longer possible to apply a custom CSS class to either the `Documents` or `Document` component.

Commit:
Responding to comments pt.3

Commit:
remove filesize input field, changing excel file indicator colour, reverting build version.

Commit:
backoffice template preview edits (WIP)

Commit:
Merging develop into this branch

Commit:
Merge branch 'develop' into feature/206838-Support-lists-of-documents

Commit:
updating example app and documentation

Ethan-Lambourne committed:
Delete docs/images/tpr-documents-block.png

Ethan-Lambourne committed:
Delete docs/images/tpr-documents-inner-blocklist.png

Ethan-Lambourne committed:
Delete docs/images/tpr-documents-input-document-form.png

Ethan-Lambourne committed:
Delete docs/images/tpr-documents-on-screen-display.png

Ethan-Lambourne committed:
Delete docs/images/tpr-documents-populated-inner-blocklist.png

Ethan-Lambourne committed:
Uploading updated example images for docs

Ethan-Lambourne committed:
Update tpr-documents.md

corrected spelling mistake, fixed some wording to reflect updates, and updated code snippet.

Commit:
Added placeholders to file size and using unformatted date for umbraco backoffice template. Also did a full uSync export.

Commit:
Merge branch 'feature/206838-Support-lists-of-documents' of https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions into feature/206838-Support-lists-of-documents

Commit:
Catching up with develop, fixing merge conflict

Ethan-Lambourne commented:
@sussexrick and @villarin I'm pretty sure all of the comments on this PR have been actioned now if you'd like to double check any of the ones you have made and resolve them if you're happy with the changes :)

Commit:
inlcuding media for examples; updating class name in backoffice preview; simplifying variables in partial view,

Ethan-Lambourne committed:
Updating wording in tpr-documents.md to reflect changes

Commit:
removing unnessecary folder inclusions

Commit:
Merge branch 'feature/206838-Support-lists-of-documents' of https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions into feature/206838-Support-lists-of-documents

Commit:
fixing tag helper html generation issue

Commit:
fixing the ability to apply custom css classes to Documents and Document components

Commit:
syncing release version with develop (will increment if no more changes are requested)
